### ISSUE TITLE
Using macroed form of field dereference causes inconsistent and misleading parse errors

### ISSUE THREAD

**Issue by fpoli** [fpoli/viper-silver#175]
Depending on which of the last three lines in the file below are uncommented, one gets:

1. A parse error regarding a "}" symbol, with the . highlighted as the error location
2. A parse error regarding a "}" symbol, with the : (from :=) highlighted as the error location
3. An error stating that an expression macro was used as a statement

It doesn't seem that any of these is the correct error. 3. is closest (see also issue #177), but the macro isn't actually in statement position, as far as I can see.

```
#!scala

domain Pointers {
  function f(x:Ref) : Ref 
}

field storedInt : Int 

define deref(ptr) ptr.storedInt

method strange() 
{
  var i: Int
  var x: Ref  
  deref(x) := loc(x).storedInt // (1.)
  //deref(x) := x.storedInt // (2.)
  //deref(i) := deref(i) // (3.)
}
```

[Issue created by alexander_summers: 2016-10-20]
[Last updated on bitbucket: 2019-02-01]

[Comment created by mschwerhoff: 2017-09-24]
See also issues #177, #200, #214

[Comment created by fabiopakk: 2019-02-01]
This issue was fixed in the following branches:

* fabiopakk_issue_177_macro_on_lhs_of_assignment
* fabiopakk_issue_178_field_deref_in_macros

They were already merged into default branch.

Errors are reported now:

1. identifier loc not defined.
2. Assignment might fail. There might be insufficient permission to access x.storedInt.
3. Type error in the expression at ...

**Comment by fpoli** [comment]
> **@mschwerhoff** commented on 2017-08-25 14:44

Resolved by @mousam05 with pull request #73

**Comment by fpoli** [comment]
> **@vakaras** commented on 2018-05-30 12:46

The following code snippet causes both verifiers to crash:

```
#!silver
field val: Int

function matrix(i: Int, j: Int): Ref
   
method test()
{
    inhale forall i: Int :: (forall j: Int :: acc(matrix(i, j).val))
}
```

Credit for finding this bug goes to Luca Campanella.

**Comment by fpoli** [comment]
> **@alexander_summers** commented on 2018-10-03 16:10

See also issue #246. We should still add the consistency check, since ASTs built with the API may rely on it.

**Comment by fpoli** [comment]
> **@alexander_summers** commented on 2018-10-08 15:38

Fixed in commit 89e1cb127f68

**Comment by fpoli** [comment]
> **@alexander_summers** commented on 2018-10-08 15:38

The fix supports this case, rather than ruling it out
